### PR TITLE
Implement the init command for user config creation  (fix #33)

### LINKED ISSUE #33: Create a config file with the init command
# User Story

As a user I want to have a command which creates a user config file, so i can store global information and the user isn`t required to enter the same data every time 

# Acceptance Criteria 

1. This file should contain the following data: 

        name = "NAME"
        team  = "TEAM"
        current_day = "DATE"
        start_year =  "START-YEAR"

2. Given the user calls the init command  and   the config file already exists,  then the old file should not be overwritten
3. Given the user calls the init command and a  config file already exist,  then the script outputs information message
4. Given the user calls the init command , when no config file exists the config file is created like `reportdaily  init --name "NAME"  --team "TEAM"   --year "START_YEAR" `
5. Given the init command is implemented, then the manpage contain an entry for the init command and its options. The README should contain a short notice about the init command.
6. Given the user calls the init subcommand without any options, then the script should ask for direct input of the required data. Just in the case the config file doesnt exist.
7. Given the init was used and all the required date was stored , then a config file appears  at  `~/.config/reportdaily/reportdailyrc`
8. GIven the user want to change his configuration, then there should be a -c and a --change option for this task.

##  Possible Solution

We need a config file with some default configuration for our new function.
A config parser module implementation could solve the problem
https://docs.python.org/3/library/configparser.html

### PR DESCRIPTION
**Pull request by SchleichsSalaticus**

### CODE REVIEW

File: docs/reportdaily.rst [hunk 1]
```diff
@@ -45,6 +45,16 @@ Subcommands
 
 .. HINT: Sort the subcommands alphabetically
 
+init
+~~~~
+
+Is required for the first time use of the programm
+User can enter his data directly or he need to answer some questions provided by this command
```
Comment (tomschr) [inline docs/reportdaily.rst:52]: I'd suggest to make it parallel with the other (sub)commands and starts with a verb:

```suggestion
Initializes global user data (required for the first time).
Either user data can be entered directly via options or user will be asked.
```

File: bin/reportdaily.py [hunk 2]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+    print(f"The file was created at this path {CONIFGPATH}")
+    show_config()
+
+
+def show_config():
+    """Show the configs to the user"""
+
+    # read the config
+    parser = ConfigParser()
+    parser.read(f"{CONIFGPATH}")
```
Comment (tomschr) [inline bin/reportdaily.py:124]: Use the variable directly:

```suggestion
    parser.read(CONIFGPATH)
```

File: bin/reportdaily.py [hunk 3]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
 log.addHandler(logging.NullHandler())
 
 
+def cmd_init(args):
+    """Creates an initial config file for a user config file"""
```
Comment (tomschr) [inline bin/reportdaily.py:68]: Describe the argument and its datatype.

File: bin/reportdaily.py [hunk 4]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+    # create config by hand
+    check_if_config_exists(args)
+
+    log.debug("INIT selected %s", args)
+    print("Init selected", args)
+    return 0
+
+
+def create_config(args):
+    """This function will create a config file where the user data is stored"""
```
Comment (tomschr) [inline bin/reportdaily.py:79]: Describe the argument and its datatype.

File: bin/reportdaily.py [hunk 5]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+        team = args.team
+    # year
+    if args.year == None:
+        year = int(input("What year of training are you in? "))
+    else:
+        year = args.year
+    # time
+    today_date = date.today()
+    print(
+        f"You are ready to go.  Entries for this {today_date} can  bee added now with reportdaily add")
```
Comment (tomschr) [inline bin/reportdaily.py:101]: * If you check for `None`, don't use the `==` operator. Use `is`.
* Don't use "Please" in your text. It just makes the whole string longer.
* In this case, you don't necessarily need the else part.

If I combine all these issues above, I would suggest:

```suggestion
    # name
    name = args.name
    if name is None:
        name = input("Enter your full name (firstname and surename): ")

    # team
    team = args.team
    if team is None:
        team = input("Please enter your team name: ")
        
    # year
    year = args.year
    if year == None:
        year = int(input("What year of training are you in? "))
        
    # time
    today_date = date.today()
    print(
        f"You are ready to go.  Entries for this {today_date} can  bee added now with reportdaily add")
```

File: bin/reportdaily.py [hunk 6]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+
+    # store configfiles in list
+    config_data = [name, team, today_date, year]
```
Comment (tomschr) [inline bin/reportdaily.py:104]: What's this? It seems, this list isn't used in this function, right?

File: bin/reportdaily.py [hunk 7]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+    # create a config file
+    config = ConfigParser()
+    config["settings"] = {'name': name,
+                          'team': team, 'current_day': today_date, 'start_year': year}
+
+    # create a config file
+    os.makedirs(os.path.dirname(CONIFGPATH), exist_ok=True)
+    with open(CONIFGPATH, 'w') as user_config:
+        config.write(user_config)
+    print(f"The file was created at this path {CONIFGPATH}")
```
Comment (tomschr) [inline bin/reportdaily.py:115]: Which file?

```suggestion
    print(f"The config file was created at {CONIFGPATH}")
```

File: bin/reportdaily.py [hunk 8]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+
+def show_config():
+    """Show the configs to the user"""
+
+    # read the config
+    parser = ConfigParser()
+    parser.read(f"{CONIFGPATH}")
+
+    print("Your current configuration at the moment")
+    print(f"""
```
Comment (tomschr) [inline bin/reportdaily.py:127]: You should combine them:

```suggestion
    print(f"""Your current configuration at the moment
```

File: bin/reportdaily.py [hunk 9]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+
+    try:
+        with open(f'{CONIFGPATH}') as user_config:
+            print("Configs already exists")
+            show_config()
+            if args.change == True:
+                change_config(args)
+    except FileNotFoundError:
+        print("Config File does not exist")
+        create_config(args)
```
[the remark this answers is not in the document]
Comment (tomschr) [reply bin/reportdaily.py:152]: On a second thought, maybe you don't need this function at all. Maybe move it into `create_config` work?

File: bin/reportdaily.py [hunk 10]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
+            keep_going = False
+        else:
+            print("No key in config found --> Try again")
+            keep_going = True
+
+    # add config parser to file
+    config = ConfigParser()
+    config.read(f"{CONIFGPATH}")
+    config.set("settings", f"{tmp_input}", f"{overwrite_input}")
+    with open(f"{CONIFGPATH}", "w") as configfile:
```
Comment (tomschr) [inline bin/reportdaily.py:197]: Use the variable directly:

```suggestion
    with open(CONIFGPATH, "w") as configfile:
```

File: README.md [hunk 11]
```diff
@@ -31,6 +31,7 @@ Just clone the repository and cd into the directory
 
 ### Subcommands
 
+- `init` : initialize and create config on first time
 - `new` : new day for entries
 - `add` : creates a new entry
 - `change` :change an existing entry
```
Comment (tomschr) [inline README.md:37]: Colons:

```suggestion
- `init`: initialize and create config on first time
- `new`: new day for entries
- `add`: creates a new entry
- `change`: change an existing entry
```

File: bin/reportdaily.py [hunk 12]
```diff
@@ -120,6 +264,19 @@ def parsecli(cliargs=None) -> argparse.Namespace:
...
+        'init', help="Create an initial Configuration file")
+    parser_init.set_defaults(func=cmd_init)
+    parser_init.add_argument(
+        '--name', "-n", help='This is a option for the user to enter his name direct via console')
+    parser_init.add_argument(
+        '--year', "-y", help='This is a option for the user to enters his teaching year direct via console')
+    parser_init.add_argument(
+        '--team', "-t", help='This is a option for the user to enter his team year direct via console')
+    parser_init.add_argument(
+        '--change', '-c', action='store_true', help='This is a option for the user to change an existing configuration')
```
Comment (tomschr) [inline bin/reportdaily.py:278]: It's a bit hard to follow if the help starts with "This is a option [for the user]...". Of course it is! :wink: Also "via console" is stating the obvious.

So in the first case, I'd remove all the obvious things and reduce it to the bare minimum:

    User name

Shorten the other help options in a similar way.

File: bin/reportdaily.py [hunk 13]
```diff
@@ -56,6 +64,142 @@ class MissingSubCommand(ValueError):
...
 
 
+def cmd_init(args):
+    """Creates an initial config file for a user config file"""
+
+    # create config by hand
+    check_if_config_exists(args)
+
+    log.debug("INIT selected %s", args)
+    print("Init selected", args)
```
Comment (tomschr) [inline bin/reportdaily.py:74]: You don't need the `print` as you already have a `log.debug`. Furthermore, I'd recommend to move the `log.debug` line as a first command. If there is an error before, you won't see no debug message.

**Review by tomschr — CHANGES_REQUESTED** [review]
@SchleichsSalaticus That PR needs some more cleanup, :wink: but the overall idea is good.

I'm not so sure about the `--change` option and how it works. Can you use `reportdaily init --change --name "Tux"` and it changes only the name, but leaves the others untouched?

Maybe you need to show that to me tomorrow.

File: bin/reportdaily.py [hunk 14]
```diff
@@ -1,15 +1,24 @@
 #!/usr/bin/env python3
 
+
 import argparse
 import logging
 from logging.config import dictConfig
+from pickle import TRUE
```
Comment (tomschr) [inline bin/reportdaily.py:7]: Please don't do that. It's confusing as you have `True` (builtin) and `TRUE` now. In this case I would really recommend to use the longer, but more obvious form `import pickle` and then `pickle.TRUE`.

After looking through your source code, it seems you don't use it. So better remove it:

```suggestion
```

File: bin/reportdaily.py [hunk 15]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+    Date: {parser.get("settings", "current_day")}
+    Year: {parser.get("settings", "start_year")}
+
+    If you desire to make changes to the configuration try the -c or --change option for the init command
+    """
+    )
+
+
+def create_config(args, CONFIGPATH):
+    """This function will create a config file where the user data is stored as a dict"""
```
Comment (tomschr) [inline bin/reportdaily.py:107]: Describe the arguments in the docstring.

File: bin/reportdaily.py [hunk 16]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+            return 1
+    # create a config if there is none
+    else:
+        create_config(args, CONFIGPATH)
+        show_config(CONFIGPATH)
+        return 0
+
+
+def show_config(CONFIGPATH):
+    """Show the configs to the user"""
```
Comment (tomschr) [inline bin/reportdaily.py:87]: Describe the argument in the docstring.

File: bin/reportdaily.py [hunk 17]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+
+    # create a config file
+    os.makedirs(os.path.dirname(CONFIGPATH), exist_ok=True)
+    with open(CONFIGPATH, 'w') as user_config:
+        config.write(user_config)
+    print(f"The file was created at this path {CONFIGPATH}")
+
+
+def ask_for_input(var, message):
+    if var is None:
```
Comment (tomschr) [inline bin/reportdaily.py:138]: Describe the arguments in the docstring.

File: bin/reportdaily.py [hunk 18]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+        var = input(message)
+    return var
+
+
+def how_to_change_config(args, CONFIGPATH):
+    """User can change or overwrite the configs via direct input or console"""
```
Comment (tomschr) [inline bin/reportdaily.py:144]: Describe the arguments in the docstring.

File: bin/reportdaily.py [hunk 19]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+    else:
+        namespace_config_change(args, CONFIGPATH)
+
+        result_value = 1
+    show_config(CONFIGPATH)
+    return result_value
+
+
+def namespace_config_change(args, CONFIGPATH):
+    """Input changes direct via the console, with just filling the namespace"""
```
Comment (tomschr) [inline bin/reportdaily.py:157]: Describe the arguments in the docstring.

File: bin/reportdaily.py [hunk 20]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+            config.set("settings", "name", name)
+        if team is not None:
+            config.set("settings", "team", team)
+        if year is not None:
+            config.set("settings", "start_year", year)
+
+        with open(CONFIGPATH, "w") as configfile:
+            config.write(configfile)
+        # show config to the user , so changes are visible to the user
+    print("namespace_config was selected")
```
Comment (tomschr) [inline bin/reportdaily.py:179]: Use your logging module instead `print`. For example:

```suggestion
    log.debug("namespace_config was selected")
```

File: bin/reportdaily.py [hunk 21]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+
+
+def user_input_change(args, CONFIGPATH):
+    # all user options
```
Comment (tomschr) [inline bin/reportdaily.py:183]: Describe the arguments in the docstring.

File: bin/reportdaily.py [hunk 22]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+        Your options are
+        Name
+        Team
+        Year
+        """)
+
+    # check for right user input
+    while(True):
+        tmp_input = input("Name, Team, Year? ")
+        if tmp_input in choice_table:
```
Comment (tomschr) [inline bin/reportdaily.py:201]: Hmn, what happens when the user enters a lowercase(!) `name`? Or enter `NAME`? That will not be found in the `choice_table`. As such, the if clause will fail.

To avoid such cases, use the string method `.capitalize()`:

```suggestion
        tmp_input = input("Name, Team, Year? ").capitalize()
        if tmp_input in choice_table:
```

File: bin/reportdaily.py [hunk 23]
```diff
@@ -56,6 +65,163 @@ class MissingSubCommand(ValueError):
...
+            # need to map the keys  right to the settings --> from Name to name
+            if tmp_input == "Name":
+                tmp_input = "name"
+            elif tmp_input == "Team":
+                tmp_input = "team"
+            elif tmp_input == "Year":
+                tmp_input = "start_year"
+
+            overwrite_input = input("Enter the change ")
```
Comment (tomschr) [inline bin/reportdaily.py:210]: Hmn... I think you miss something here. What happens if the user wants to change the year, but enters `foo`. Does your code cover this case?

File: tests/test_init.py [hunk 24]
```diff
@@ -0,0 +1,412 @@
+# import configparser
+import configparser
+from pickle import TRUE
```
Comment (tomschr) [inline tests/test_init.py:3]: You don't use it, so better remove it:

```suggestion
```

File: tests/test_init.py [hunk 25]
```diff
@@ -0,0 +1,412 @@
...
+import builtins
+# required fo tests
+import pytest
+
+
+# Global data
+STANDARD_SECTIONS = ["settings"]
+STANDARD_OPTIONS = ["name", "team", "start_year", "current_day"]
+STANDARD_CONFIG = {  # Section_name: list of key names
+    "settings": ("name", "team", "start_year", "current_day"),
```
Comment (tomschr) [inline tests/test_init.py:27]: Be DRY ("don't repeat yourself") here:

```suggestion
    "settings": tuple(STANDARD_OPTIONS),
```

**Review by tomschr — CHANGES_REQUESTED** [review]
@SchleichsSalaticus Most comments are in the script, not in the tests. Shouldn't be too much.

Good work! :+1:

File: bin/reportdaily.py (hunk 14 above)
Comment (SchleichsSalaticus) [reply bin/reportdaily.py:7]: Hmm strange, i need to check my vs code settings,  it seems like sometimes vs codes auto adds imports if something is missing.
I need to disable this feature. 
I cant remeber to import "True" 
Thanks for the tipp.

File: bin/reportdaily.py (hunk 22 above)
Comment (SchleichsSalaticus) [reply bin/reportdaily.py:201]: Oh Thanks, i knew there is a way , but didnt now about capitalize()
Thanks

File: tests/test_init.py (hunk 24 above)
Comment (SchleichsSalaticus) [reply tests/test_init.py:3]: I will delete this import

File: tests/test_init.py (hunk 25 above)
Comment (SchleichsSalaticus) [reply tests/test_init.py:27]: Thanks your are right  :dancers:

File: bin/reportdaily.py [hunk 26]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+    If you desire to make changes to the configuration try the -c or --change option for the init command
+    """
+    )
+
+
+def create_config(args, CONFIGPATH):
+    """
+    Create a config file, from users input, where the user data is stored as a dict
+    :param  namespace: Arguments given by the command line
+    :param  str: Path where the configuration file is stored
```
Comment (tomschr) [inline bin/reportdaily.py:115]:
```suggestion
    :param argparser.Namespace args: Arguments given by the command line
    :param str CONFIGPATH: Path where the configuration file is stored
```

File: bin/reportdaily.py [hunk 27]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
 log.addHandler(logging.NullHandler())
 
 
+def cmd_init(args, CONFIGPATH):
+    """
+    Initializes global user data (required for the first time).
+    Either user data can be entered directly via options or user will be asked.
+    :param  namespace: Arguments given by the command line
+    :param  str: Path where the configuration file is stored
+    :rtype: int for tests
```
Comment (tomschr) [inline bin/reportdaily.py:70]:
```suggestion
    Either user data can be entered directly via options or user will be asked.

    :param argparse.Namespace args: Arguments given by the command line
    :param str CONFIGPATH: Path where the configuration file is stored
    :return: exit code of this function
    :rtype: int
```

File: bin/reportdaily.py [hunk 28]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+    else:
+        create_config(args, CONFIGPATH)
+        show_config(CONFIGPATH)
+        return 0
+
+
+def show_config(CONFIGPATH):
+    """
+    Show the configs to the user
+    :param str: String with an absolute path for looking up the values of the configfile
```
Comment (tomschr) [inline bin/reportdaily.py:91]:
```suggestion
    Show the configs to the user

    :param str CONFIGPATH: String with an absolute path for looking up the values of the configfile
```

File: bin/reportdaily.py [hunk 29]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+        var = input(message)
+    return var
+
+
+def how_to_change_config(args, CONFIGPATH):
+    """
+    Change or overwrite the configs via direct input or cli Attributes
+    :param  namespace: Attributes given by the command line
+    :param  str: Path where the configuration file is stored
+    :return int: int return value for testing
```
Comment (tomschr) [inline bin/reportdaily.py:156]:
```suggestion
    Change or overwrite the configs via direct input or cli Attributes

    :param argparse.Namespace args: Attributes given by the command line
    :param str CONFIGPATH: Path where the configuration file is stored
    :return int: int return value for testing
```

File: bin/reportdaily.py [hunk 30]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+        result_value = 1
+    show_config(CONFIGPATH)
+    return result_value
+
+
+def namespace_config_change(args, CONFIGPATH):
+    """
+    Input arguments direct via the console
+    :param  namespace: Attributes given by the command line
+    :param  str: Path where the configuration file is stored
```
Comment (tomschr) [inline bin/reportdaily.py:173]:
```suggestion
    Input arguments direct via the console

    :param argparse.Namespace args: Attributes given by the command line
    :param str CONFIGPATH: Path where the configuration file is stored
```

File: bin/reportdaily.py [hunk 31]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+        # show config to the user , so changes are visible to the user
+    log.debug("namespace_config was selected")
+
+
+def check_is_int(input_str, input_is_int):
+    """
+    Prove if the given argument is an int and return True or decline and return False
+    :param  str: String given that needs to be checked
+    :param  bool: bool value default False
+    :return bool value, True if str is an int, False if str is not an int
```
Comment (tomschr) [inline bin/reportdaily.py:204]:
```suggestion
    Prove if the given argument is an int and return True or decline and return False

    :param str input_str: String given that needs to be checked
    :param bool input_is_int: bool value default False
    :return: True if str is an int, False if str is not an int
    :rtype: bool
```

File: bin/reportdaily.py [hunk 32]
```diff
@@ -56,6 +61,216 @@ class MissingSubCommand(ValueError):
...
+        print("Input is not a int sorry, try again")
+        input_is_int = False
+        return input_is_int
+
+
+def user_input_change(args, CONFIGPATH):
+    """
+    Input the data that is asked by function tp change configs
+    :param  str: String given that needs to be checked
+    :param  bool: bool value default False
```
Comment (tomschr) [inline bin/reportdaily.py:221]:
```suggestion
    Input the data that is asked by function tp change configs

    :param argparse.Namespace args: Attributes given by the command line
    :param str CONFIGPATH: Path where the configuration file is stored
    :return: the ConfigParser object
    :rtype: configparse.ConfigParser
```

File: bin/reportdaily.py [hunk 33]
```diff
@@ -99,10 +314,10 @@ def cmd_export(args):
 
 
 def parsecli(cliargs=None) -> argparse.Namespace:
-    """Parse CLI with :class:`argparse.ArgumentParser` and return parsed result
+    """Parse CLI with: class: `argparse.ArgumentParser` and return parsed result
```
Comment (tomschr) [inline bin/reportdaily.py:317]: The space between `:class:` and the class name is essential. It's actually a macro which doesn't allow any whitespaces in between:

```suggestion
    """Parse CLI with: class:`argparse.ArgumentParser` and return parsed result
```

**Review by tomschr — CHANGES_REQUESTED** [review]
@SchleichsSalaticus Looks quite good! :+1: 

There are only minor things, for example, an empty line between purpose of the function and its parameters. Also the description of parameters needs the syntax `:param <TYPE> <NAME>`, you missed the name.

I think, it's quite easy this time: just accept my suggestions and you are done. :-)
